AM62l lpm clk and USB fix

ti-sebin opened the pull request.

ti-sebin committed:
PENDING: fix(ti): Update the clock mux driver

Update the clock divider set function to write to the divider register
only when the writing value doesn't match with the current reg value.
This is done to avoid unnecessary write to dividers that are protected by
firewalls

Change-Id: I5d3733ccbe27c95aac8882b13473560191dbfaf7
Signed-off-by: Sebin Francis <sebin.francis@ti.com>

ti-sebin committed:
PENDING: feat(ti): suspend and restore clocks during lpm

Clock configurations are lost during LPM entry. Save and restore
the clock configurations during LPM.

Change-Id: I051595536819c60294a8979b69b7ba789dfc5d46
Signed-off-by: Sebin Francis <sebin.francis@ti.com>

DhruvaG2000 commented on plat/ti/k3/board/am62l/board.mk:
```diff
@@ -12,6 +12,8 @@ DDR_16BIT	:=	1
 K3_AM62L_LPM	:=	1
 $(eval $(call add_define,K3_AM62L_LPM))
+CONFIG_LPM_CLK	:=	1
+$(eval $(call add_define,CONFIG_LPM_CLK))
```
I am just curious, in what scenario would we not need this? Why even make it a CONFIG?

akashdeep-ti replied:
yes, will there be any scenario where customers will not like to restore clk configuration. If no, then, we can replace this in clk file with LPM macro itself

ti-sebin replied:
agreed, will remove the config

DhruvaG2000 commented on plat/ti/k3/common/drivers/pm/clock/clk_div.c:
```diff
@@ -464,17 +464,19 @@ bool clk_div_reg_set_div(struct clk *clkp, uint32_t d)
 		v |= d_val_p << data_reg->bit;
-		ti_clk_writel(v, (uint32_t) data_reg->reg);
+		if(v != reg) {
+			ti_clk_writel(v, (uint32_t) data_reg->reg);
```
in any case, if we know this is protected, won't it cause an exception irrespective of what the value is?

ti-sebin replied:
No, these protected clock will be restored by TIFS prior to this step so the value will always match and there will not be any write required to this register.

DhruvaG2000 replied:
Can we please add this explanation as a comment? I don't think any commit message mentions about TIFS restoring the clocks either.

ti-sebin replied:
Update the PR

akashdeep-ti commented on plat/ti/k3/common/am62l_psci.c:
```diff
@@ -141,6 +142,7 @@ static void am62l_pwr_domain_suspend(const psci_power_state_t *target_state)
 	/* Prevent interrupts from spuriously waking up this cpu */
 	k3_gic_cpuif_disable();
 	k3_gic_save_context();
+	clks_suspend();
```
If the system fails to suspend because of incorrect mode, will the clks be resumed?

ti-sebin replied:
Once the code reaches this point, we cant return back to Linux if there is a failure. Only option is to continue the lpm sequence and try to resume.

akashdeep-ti commented:
Minor comments, otherwise the PR looks good to me

ti-sebin committed:
PENDING: feat(ti): Add suspend and resume function to clk driver

Add suspend and resume callbacks to the clock driver. This will be
used in LPM to save and restore clocks.

Change-Id: I12e717d9aca38e00e1b6e72c48131dffa8500264
Signed-off-by: Sebin Francis <sebin.francis@ti.com>

ti-sebin committed:
PENDING: fix(ti): Update USB LPSC flags

Remove the `LPSC_NO_CLOCK_GATING` flag from USB's ISO LPSCs as it
was preventing the ISO LPSC to be turned off, when USB is turned
off. Also these LPSC doesn't required to be allways ON. Its needs
to be only on when USB is in use.

Change-Id: I97d0f80f4a6447621916a96c1fc2be7459ee7427
Signed-off-by: Sebin Francis <sebin.francis@ti.com>

DhruvaG2000 reviewed (changes requested):
Can you run checkpatch on all the patches please?
```
dhruva:trusted-firmware-a/ (am62l_lpm_clk_fix✗) $ /home/dhruva/my_workspace/kernels/clean/linux-next/scripts/checkpatch.pl -g HEAD~2 --strict                                                           [11:28:28]
plat/ti/k3/common/drivers/pm/clock/clk.c:643: CHECK: No space is necessary after a cast
+               parent_clk = clk_lookup((clk_idx_t) p->clk);

plat/ti/k3/common/drivers/pm/clock/clk.c:682: WARNING: Missing a blank line after declarations
+                       struct clk *clkp = soc_clocks + i;
+                       ret = clk_suspend_save(clkp);

plat/ti/k3/common/drivers/pm/clock/clk_mux.c:100: CHECK: Using comparison to false is error prone
+       if (error == false) {

total: 0 errors, 1 warnings, 2 checks, 283 lines checked
dhruva:trusted-firmware-a/ (am62l_lpm_clk_fix✗) $ /home/dhruva/my_workspace/kernels/clean/linux-next/scripts/checkpatch.pl -g HEAD~3 --strict                                                           [11:28:35]
plat/ti/k3/common/drivers/pm/clock/clk_div.c:475: CHECK: No space is necessary after a cast
+               v = reg & ((uint32_t) ~(((1U << (uint32_t) ilog32(n)) - 1U) << data_reg->bit));

plat/ti/k3/common/drivers/pm/clock/clk_div.c:478: ERROR: space required before the open parenthesis '('
+               if(v != reg) {

plat/ti/k3/common/drivers/pm/clock/clk_div.c:479: CHECK: No space is necessary after a cast
+                       ti_clk_writel(v, (uint32_t) data_reg->reg);

plat/ti/k3/common/drivers/pm/clock/clk_div.c:480: ERROR: trailing whitespace
+^I^I}^I^I$

total: 2 errors, 0 warnings, 2 checks, 24 lines checked
```

ti-sebin commented:
> Can you run checkpatch on all the patches please?
> 
> ```
> dhruva:trusted-firmware-a/ (am62l_lpm_clk_fix✗) $ /home/dhruva/my_workspace/kernels/clean/linux-next/scripts/checkpatch.pl -g HEAD~2 --strict                                                           [11:28:28]
> plat/ti/k3/common/drivers/pm/clock/clk.c:643: CHECK: No space is necessary after a cast
> +               parent_clk = clk_lookup((clk_idx_t) p->clk);
> 
> plat/ti/k3/common/drivers/pm/clock/clk.c:682: WARNING: Missing a blank line after declarations
> +                       struct clk *clkp = soc_clocks + i;
> +                       ret = clk_suspend_save(clkp);
> 
> plat/ti/k3/common/drivers/pm/clock/clk_mux.c:100: CHECK: Using comparison to false is error prone
> +       if (error == false) {
> 
> total: 0 errors, 1 warnings, 2 checks, 283 lines checked
> dhruva:trusted-firmware-a/ (am62l_lpm_clk_fix✗) $ /home/dhruva/my_workspace/kernels/clean/linux-next/scripts/checkpatch.pl -g HEAD~3 --strict                                                           [11:28:35]
> plat/ti/k3/common/drivers/pm/clock/clk_div.c:475: CHECK: No space is necessary after a cast
> +               v = reg & ((uint32_t) ~(((1U << (uint32_t) ilog32(n)) - 1U) << data_reg->bit));
> 
> plat/ti/k3/common/drivers/pm/clock/clk_div.c:478: ERROR: space required before the open parenthesis '('
> +               if(v != reg) {
> 
> plat/ti/k3/common/drivers/pm/clock/clk_div.c:479: CHECK: No space is necessary after a cast
> +                       ti_clk_writel(v, (uint32_t) data_reg->reg);
> 
> plat/ti/k3/common/drivers/pm/clock/clk_div.c:480: ERROR: trailing whitespace
> +^I^I}^I^I$
> 
> total: 2 errors, 0 warnings, 2 checks, 24 lines checked
> ```

Thanks for the review, Update the PR with the fixes

akashdeep-ti commented on plat/ti/k3/common/drivers/pm/clock/clk_div.c:
```diff
@@ -464,17 +464,20 @@ bool clk_div_reg_set_div(struct clk *clkp, uint32_t d)
+		v = reg & ((uint32_t)~(((1U << (uint32_t)ilog32(n)) - 1U) << data_reg->bit));
 		v |= d_val_p << data_reg->bit;
-		ti_clk_writel(v, (uint32_t) data_reg->reg);
+		/* Update the register only if value does,t match */
```
/* Update the register only if value doesn't match */

ti-sebin replied:
Thanks Updated

akashdeep-ti commented on plat/ti/k3/common/drivers/pm/clock/clk.c:
```diff
@@ -607,6 +607,141 @@ int32_t clk_deinit_pm_devgrp(uint8_t pm_devgrp)
+				clkp->flags |= CLK_FLAG_SUSPENDED;
+			} else {
+				/* Mark clock as resumed if no handler is provided */
+				clkp->flags |= CLK_FLAG_SUSPENDED;
```
the comment says to mark clock as resumed but in flags, we are setting as suspended.
Should it not be clkp->flags &= ~CLK_FLAG_SUSPENDED;? or the comment needs update

ti-sebin replied:
Good catch, comment needs update, updated the comment

akashdeep-ti commented on plat/ti/k3/common/drivers/pm/clock/clk.c:
```diff
@@ -607,6 +607,141 @@ int32_t clk_deinit_pm_devgrp(uint8_t pm_devgrp)
+	uint32_t i;
+	int32_t ret = SUCCESS;
+	uint32_t clock_count = soc_clock_count;
+	uint8_t max_tries = 10;
```
can we use the macro LPM_CLK_MAX_TRIES?

ti-sebin replied:
Yes, Updated the PR
